### PR TITLE
Support foreach over parameters in DoNotCopyValue

### PR DESCRIPTION
**Pull request by sharwell**
Builds on #4368

### CODE REVIEW

File: src/Roslyn.Diagnostics.Analyzers/Core/DoNotCopyValue.cs [hunk 1]
```diff
@@ -9,6 +9,8 @@
 using Analyzer.Utilities;
 using Analyzer.Utilities.Extensions;
 using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.CSharp;
+using Microsoft.CodeAnalysis.CSharp.Syntax;
```
Comment (mavasani) [inline src/Roslyn.Diagnostics.Analyzers/Core/DoNotCopyValue.cs:13]: Not recommended in the shared layer. Please move to the C# layer.

File: src/Roslyn.Diagnostics.Analyzers/Core/DoNotCopyValue.cs [hunk 2]
```diff
@@ -547,14 +549,36 @@ public override void VisitForEachLoop(IForEachLoopOperation operation)
...
+
+                switch (Acquire(operation.Collection))
                 {
-                    instance = null;
-                    instance2 = null;
+                    case RefKind.Ref:
+                        // No special requirements
+                        break;
+
+                    case RefKind.RefReadOnly when operation.Syntax is CommonForEachStatementSyntax syntax && operation.SemanticModel.GetForEachStatementInfo(syntax).GetEnumeratorMethod is { IsReadOnly: true }:
```
Comment (mavasani) [inline src/Roslyn.Diagnostics.Analyzers/Core/DoNotCopyValue.cs:559]: This is getting too difficult to manage - IMO this analyzer is not sustainable unless rewritten as a flow-based analyzer. I'll let you decide how to take this forward.

**Review by mavasani — APPROVED** [review]
At this point, I am not sure my reviews are valuable for this analyzer :-) - the analyzer feels way too complex and syntax/language construct dependent to me.

File: src/Roslyn.Diagnostics.Analyzers/Core/DoNotCopyValue.cs (hunk 2 above)
Comment (sharwell) [reply src/Roslyn.Diagnostics.Analyzers/Core/DoNotCopyValue.cs]: > IMO this analyzer is not sustainable unless rewritten as a flow-based analyzer.

Flow based analysis does not help here. Every language construct needs to be allow-listed as it is encountered.

File: src/Roslyn.Diagnostics.Analyzers/Core/DoNotCopyValue.cs (hunk 1 above)
Comment (sharwell) [reply src/Roslyn.Diagnostics.Analyzers/Core/DoNotCopyValue.cs:13]: Is there a language independent way to get the `GetEnumerator` method used by the `IForEachLoopOperation`?

Comment (sharwell) [reply src/Roslyn.Diagnostics.Analyzers/Core/DoNotCopyValue.cs]: ➡️ Moved language-specific functionality to derived types

**Pull request merged**